plugin/helpers_kv.go

```diff
@@ -129,3 +131,72 @@ func (p *HelpersImpl) KVSetWithExpiryJSON(key string, value interface{}, expireI
 
 	return nil
 }
+
+// KVAtomicModify is a wrapper around KVGet and KVCompareAndSet that atomically modifies data for a particular key. This function
+// takes the following parameters:
+// - ctx 	An instance of a request context that should be used for stopping the transaction in the case of downstream
+// 			cancellations.
+// - key	A key string for used querying the data to be modified.
+// - bucket	An instance of a token bucket that should be used for throttling control on retries.
+// - fn 	A function that modifies the data queried for the specific key.
+func (p *HelpersImpl) KVAtomicModify(ctx context.Context, key string, bucket einterfaces.TokenBucket, fn func(initialValue []byte) ([]byte, error)) error {
```

Comment:
Could you elaborate a bit why do choose this complex interface?
1. For what use case ist a `Context` needed? I can think of a case where you want to cancel such operation. We also don't use `Context` in any other API or helper method.
2. Why does the plugin dev need such control over the throttling via a `TokenBucket`? To me throttling should be handelt my the helper implementation and not by the caller. Also, I can't think of a use case where the plugin dev would care about this.

Reply:
@gsagula: @hanzei's comments mirror mine in our offline conversation. I think we can significantly simplify at the cost of some lost functionality that is not immediately useful. Specifically, I too questioned the need for the Context, and the abstract TokenBucket throttling implementation (I was in favor of a simple, limited retry loop followed by a failure.

For more context, see one of my favorite "blog posts" of all time, https://www.jwz.org/doc/worse-is-better.html.

Reply:
@levb @hanzei Thanks for the review.

> For what use case ist a Context needed? I can think of a case where you want to cancel such operation. We also don't use Context in any other API or helper method.

I'm not sure if it is to critical in the case of the RPC plugin, but it's fundamentally a good practice to always provide a context with a deadline on downstream requests. For instance, if anything happens during the HTTP request produced by the command, database transactions or other related operations should also get cancelled or reverted so users don't end up with a responses that do not reflect the state of the system. In the case of AtomicModify, we have a get transaction followed by a compare and modify, both being called interactively in a loop. Another good reason for passing context throughout the  application in early development stages is that it will be much easier for rolling out tracing, stats and other good things later on, if desirable. This last one is totally based on my experience though.

> Why does the plugin dev need such control over the throttling via a TokenBucket? To me throttling should be handelt my the helper implementation and not by the caller. Also, I can't think of a use case where the plugin dev would care about this

That's a good point. I agree that the bucket or any other algorithm used to protect the system from exhaustive operations would be better if handled internally by the helper.

@levb Thanks for the article. I'm not opposed to simplify the design by limiting the retries with a hardcoded value, however I think that we will need some visibility over this operation in case it starting to fall often in production. Regarding context, please see if my explanation above makes sense. That was kind of what I was thinking when decided to add context to this interface.

Reply:
Sorry for the late reply.

Just to be clear: The plugin API uses RPC, but HTTP to make the requests. If there is any RPC failure, there plugin process will be terminated anyway. I strongly suppose that failures will happen rarely in production.

> Another good reason for passing context throughout the application in early development stages is that it will be much easier for rolling out tracing, logging and other good things later on, if desirable.

Can you elaborate a bit please? Why would you need a `context` for this?